File: 4 sem/ADS/task19/19.py
l = int(input())
nums = list(map(int, input().split()))

inv = 0

def merge_sort(n):
    global inv
    nl = len(n) - 1

    m = nl//2

    l = n[:m+1]
    r = n[m+1:]

    ll = len(l)
    rl = len(r)

    if ll > 1:
        l = merge_sort(l)
    if rl > 1:
        r = merge_sort(r)
    
    i = 0
    j = 0
    res = []

    while i < ll and j < rl:
        if r[j] < l[i]:
            res.append(r[j])
            inv += m + 1 - i
            j += 1
        else:
            res.append(l[i])
            i += 1
    
    while i < ll:
        res.append(l[i])
        i += 1
    while j < rl:
        res.append(r[j])
        j += 1
    
    return res

merge_sort(nums)
print(inv)

File: 4_sem/ADS/19.cpp
#include <iostream>
#include <string>
#include <vector>

using std::cin;
using std::cout;
using std::string;
using std::vector;

int inv = 0;

void merge_sort(int* v, int start, int finish)
{
	int m = (finish + start) / 2;

	if ((m + 1 - start) > 1)
	{
		merge_sort(v, start, m);
	}
	if ((finish - m) > 1)
	{
		merge_sort(v, m + 1, finish);
	}
	
	int i = start;
	int j = m + 1;
	vector<int> res;

	while ((i < m + 1) && (j <= finish))
	{
		if (v[j] < v[i])
		{
			res.push_back(v[j]);
			inv += m + 1 - i;
			j += 1;
		}
		else
		{
			res.push_back(v[i]);
			i += 1;
		}
	}
	while (i < m + 1)
	{
		res.push_back(v[i]);
		i += 1;
	}
	while (j <= finish)
	{
		res.push_back(v[j]);
		j += 1;
	}
	for (int _i = 0; _i < finish-start+1; _i++)
	{
		v[start + _i] = res[_i];
	}
	res.clear();
}

int main() 
{
	int n;
	string s;
	cin >> n;
	int* nums = new int[n];
    
	for (int i = 0; i < n; i++)
	{
		cin >> nums[i];
	}
	
	merge_sort(nums, 0, n-1);

	cout << inv;
}
	